app/src/lib/stores/helpers/repository-indicator-updater.ts

```diff
@@ -0,0 +1,168 @@
+      )
+
+      this.refreshTimeoutId = window.setTimeout(
+        () => this.refreshAllRepositories(),
+        timeout
+      )
+    }
+  }
+
+  private async refreshAllRepositories() {
+    // We're only ever called by the setTimout so it's safe for us to clear
+    // this without calling clearTimeout
+    this.refreshTimeoutId = null
+    log.debug('[RepositoryIndicatorUpdater] Running refreshAllRepositories')
+    if (this.paused) {
+      log.debug(
+        '[RepositoryIndicatorUpdater] Paused before starting refreshAllRepositories'
+      )
+      await this.pausePromise
+
+      if (!this.running) {
+        return
+      }
+    }
+
+    this.lastRefreshStartedAt = Date.now()
+
+    let repository
+    const done = new Set<number>()
+    const getNextRepository = () =>
+      this.getRepositories().find(x => !done.has(x.id))
+
+    const startTime = Date.now()
+    let pausedTime = 0
+
+    while (this.running && (repository = getNextRepository()) !== undefined) {
+      await this.refreshRepositoryIndicators(repository)
+
+      if (this.paused) {
+        log.debug(
+          `[RepositoryIndicatorUpdater] Pausing after ${done.size} repositories`
+        )
+        const pauseTimeStart = Date.now()
+        await this.pausePromise
+        pausedTime += Date.now() - pauseTimeStart
+        log.debug(
+          `[RepositoryIndicatorUpdater] Resuming after ${pausedTime / 1000}s`
+        )
+      }
+
+      done.add(repository.id)
+    }
+
+    if (done.size >= 1) {
+      const totalTime = Date.now() - startTime
+      const activeTime = totalTime - pausedTime
+      const activeTimeSeconds = (activeTime / 1000).toFixed(1)
+      const pausedTimeSeconds = (pausedTime / 1000).toFixed(1)
+      const totalTimeSeconds = (totalTime / 1000).toFixed(1)
+
+      log.info(
+        `[RepositoryIndicatorUpdater]: Refreshing sidebar indicators for ${done.size} repositories took ${activeTimeSeconds}s of which ${pausedTimeSeconds}s paused, total ${totalTimeSeconds}s`
+      )
+    }
+
+    this.scheduleRefresh()
+  }
+
+  private clearRefreshTimeout() {
+    if (this.refreshTimeoutId !== null) {
+      window.clearTimeout()
+      this.refreshTimeoutId = null
+    }
+  }
+
+  public stop() {
+    if (this.running) {
+      log.debug('[RepositoryIndicatorUpdater] Stopping')
+      this.running = false
+      this.clearRefreshTimeout()
+    }
+  }
+
+  public pause() {
+    if (this.paused === false) {
+      log.debug('[RepositoryIndicatorUpdater] Pausing')
+
+      // Disable the lint warning since we're storing the `resolve`
+      // tslint:disable-next-line:promise-must-complete
+      this.pausePromise = new Promise<void>(resolve => {
+        this.resolvePausePromise = resolve
+      })
+
+      this.paused = true
+    }
+  }
+
+  public resume() {
+    if (this.paused) {
+      log.debug('[RepositoryIndicatorUpdater] Resuming')
+
+      this.pausePromise = Promise.resolve()
```

Comment:
Why do we need to reassign `pausePromise` here to a resolved promise if we're already resolving the existing `pausePromise` promise?

Reply:
I don't think we do, I believe this was a leftover from a previous attempt of mine. Nice catch!